python/cugraph/cugraph/gnn/data_loading/dist_sampler.py

```diff
@@ -0,0 +1,349 @@
+                    "map": renumber_map_array_p,
+                    "label_hop_offsets": label_hop_offsets_array_p,
+                    "weight": weight_array_p,
+                    "edge_id": edge_id_array_p,
+                    "edge_type": edge_type_array_p,
+                    "renumber_map_offsets": renumber_map_offsets_array_p,
+                }
+            )
+
+            end_batch_id = start_batch_id + len(batch_id_array_p) - 1
+            rank = minibatch_dict["rank"] if "rank" in minibatch_dict else 0
+
+            full_output_path = os.path.join(
+                self.__directory,
+                f"batch={rank:05d}.{start_batch_id:08d}-"
+                f"{rank:05d}.{end_batch_id:08d}.parquet",
+            )
+
+            results_dataframe_p.to_parquet(
+                full_output_path,
+                compression=None,
+                index=False,
+                force_nullable_schema=True,
+            )
+
+    def __write_minibatches_csr(self, minibatch_dict):
+        raise NotImplementedError(
+            "CSR format currently not supported for distributed sampling"
+        )
+
+    def write_minibatches(self, minibatch_dict):
+        if (minibatch_dict["majors"] is not None) and (
+            minibatch_dict["minors"] is not None
+        ):
+            self.__write_minibatches_coo(minibatch_dict)
+        elif (minibatch_dict["major_offsets"] is not None) and (
+            minibatch_dict["minors"] is not None
+        ):
+            self.__write_minibatches_csr(minibatch_dict)
+        else:
+            raise ValueError("invalid columns")
+
+
+class DistSampler:
+    def __init__(
+        self,
+        graph: Union[pylibcugraph.SGGraph, pylibcugraph.MGGraph],
+        writer: DistSampleWriter,
+        local_seeds_per_call: int = 32768,
+    ):
+        self.__graph = graph
+        self.__writer = writer
+        self.__local_seeds_per_call = local_seeds_per_call
+
+    def sample_batches(
+        self, seeds: TensorType, batch_ids: TensorType, random_state: int = 0
+    ):
+        raise NotImplementedError("Must be implemented by subclass")
+
+    def sample_from_nodes(
+        self, nodes: TensorType, *, batch_size: int = 16, random_state: int = 62
+    ):
+        batches_per_call = self._local_seeds_per_call // batch_size
+        actual_seeds_per_call = batches_per_call * batch_size
+
+        num_seeds = len(nodes)
+        nodes = torch.split(
+            torch.as_tensor(nodes, device="cuda"), actual_seeds_per_call
+        )
+
+        if self.is_multi_gpu:
+            rank = torch.distributed.get_rank()
+            world_size = torch.distributed.get_world_size()
+
+            t = torch.empty((world_size,), dtype=torch.int64, device="cuda")
+            local_size = torch.tensor(
+                [int(ceil(num_seeds / batch_size))], dtype=torch.int64, device="cuda"
+            )
+
+            torch.distributed.all_gather_into_tensor(t, local_size)
```

Comment:
We need to know where the batch ids start for the rank.  Otherwise we'll get multiple ranks using the same batch ids, which will cause invalid samples to be produced.  So this all gather gets the number of batches for each rank, and figures out the starting batch id for this rank by doing a cumulative sum (line 212).

I think I may split this into a separate function.  Also I may add a way to skip this logic if you're 100% confident all ranks have the same number of batches (I have a FIXME in line 273 to do the same thing in another part of the code)

Reply:
Done.